## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [31]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [32]:
# View the column names in the data
df.columns 


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [33]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [34]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.job.unique().tolist()


['Immunologist',
 'Music therapist',
 'Accounting technician',
 'Chiropractor',
 'Production designer, theatre/television/film',
 'Barista',
 'Theme park manager',
 'Chiropodist',
 'Secretary, company',
 'Engineer, broadcasting (operations)',
 'Surveyor, land/geomatics',
 'Lobbyist',
 'Psychologist, clinical',
 'Research scientist (maths)',
 'Government social research officer',
 'Colour technologist',
 'Production assistant, television',
 'Lecturer, further education',
 'Commercial art gallery manager',
 'Architectural technologist',
 'Engineer, manufacturing systems',
 'Textile designer',
 'Air cabin crew',
 'Medical illustrator',
 'Technical brewer',
 'Radiographer, therapeutic',
 'Race relations officer',
 'Horticulturist, amenity',
 'Surveyor, rural practice',
 'Science writer',
 'Airline pilot',
 'Environmental health practitioner',
 'Warehouse manager',
 'IT technical support officer',
 'Quarry manager',
 'Accountant, chartered',
 'Doctor, general practice',
 'Structural enginee

In [35]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [36]:
# For the category with the most entries,
# which subcategory had the most entries?
df.loc[df["category"] == "consumables", "subcategory"].value_counts().head(1)



subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [37]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head(5).index


Index([33615, 66037, 46820, 38378, 24741], dtype='int64', name='client_id')

In [38]:
# Store the client ids of those top 5 clients in a list
top_client_ids = list(df["client_id"].value_counts().head(5).index)
top_client_ids


[33615, 66037, 46820, 38378, 24741]

In [39]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df["client_id"] == 33615, "qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [40]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [41]:
# Create a column for shipping price.Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df.apply(lambda row: row["unit_weight"] * row["qty"] * 7 if row["unit_weight"] * row["qty"] > 50 else row["unit_weight"] * row["qty"] * 10, axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  \
0     1096.80     762.71         7.50  105            1      115164.00   
1       24.95      15.09         1.49   21            0         523.95   
2       13.52       7.86         1.68   39            6         527.28   
3       36.42      24.85         1.23   29            3        1056.18   
4      195.10     108.17        46.43   20            1        3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [42]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = (df["line_subtotal"] + df["shipping_price"] ) * 1.0925
df.head(10)


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ...     category  \
0   8953482  2023-04-28          17        2023  ...        decor   
1   8069089  2023-05-19          20        2023  ...  consumables   
2   1902144  2023-01-29           4        2023  ...     software   
3   9031802  2023-04-25          17        2023  ...  consumables   
4   1322274  2023-05-28          21        2023  ...  consumables   
5   1327730  2023-05-02          18        2023  ...        decor   
6   8080280  2023-04-19          16        2023  ...  consumables   
7   8946782  2023-02-14           7        2023  ...  consumables   
8   9394118  2023-05-13          19        2023  ...  consumables   
9   9665852  2023-04-14          15        2023  ...    furniture   

          subcategory unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art    1096.80     762.71         7.50  105            1   
1                pens      24.95      15.09         1.49   21            0   
2  project management      13.52       7.86         1.68   39            6   
3                pens      36.42      24.85         1.23   29            3   
4                misc     195.10     108.17        46.43   20            1   
5              plants     172.84     154.44         2.79   78            0   
6    kitchen supplies     245.56     231.95         1.73   56            5   
7                pens      20.99      16.97         5.12   22            4   
8      paper products      83.86      59.00         0.27   49            3   
9             seating      66.88      54.15         2.47   78            1   

   line_subtotal  shipping_price    total_price  
0      115164.00         5512.50  131839.076250  
1         523.95          312.90     914.258625  
2         527.28          458.64    1077.117600  
3        1056.18          356.70    1543.571400  
4        3902.00         6500.20   11364.403500  
5       13481.52         1523.34   16392.809550  
6       13751.36          678.16   15764.250600  
7         461.78          788.48    1365.909050  
8        4109.14          132.30    4633.773200  
9        5216.64         1348.62    7172.546550  

[10 rows x 21 columns]

In [43]:
# Create a column for the cost of each line using unit cost, qty, and shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price  \
0         7.50  105            1      115164.00         5512.50   
1         1.49   21            0         523.95          312.90   
2         1.68   39            6         527.28          458.64   
3         1.23   29            3        1056.18          356.70   
4        46.43   20            1        3902.00         6500.20   

     total_price  line_cost  
0  131839.076250   85597.05  
1     914.258625     629.79  
2    1077.117600     765.18  
3    1543.571400    1077.35  
4   11364.403500    8663.60  

[5 rows x 22 columns]

In [44]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["total_price"] - df["line_cost"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  shipping_price    total_price  line_cost  \
0            1      115164.00         5512.50  131839.076250   85597.05   
1            0         523.95          312.90     914.258625     629.79   
2            6         527.28          458.64    1077.117600     765.18   
3            3        1056.18          356.70    1543.571400    1077.35   
4            1        3902.00         6500.20   11364.403500    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [45]:
# Check your work using the totals above
def calculate_total_price(order_id, df):
    total_price = df[df['order_id'] == order_id]['total_price'].sum()
    return total_price

order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    total_price = calculate_total_price(order_id, df)
    display(f"Order ID {order_id} has a total price of ${total_price}")


'Order ID 2742071 has a total price of $152811.897075'

'Order ID 2173913 has a total price of $162388.7193'

'Order ID 6128929 has a total price of $923441.24045'

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [46]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
def client_spend(client_id, total_price):
    client_df = df.loc[df['client_id'] == client_id, total_price]
    return round(client_df.sum(), 2)
for client_id in top_client_ids:
    print(f"{client_id}: ${client_spend(client_id, 'total_price')}")


33615: $8377308.57
66037: $10259514.8
46820: $9743794.32
38378: $12906550.88
24741: $82268891.98


In [47]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
data_dict = {"client_id": top_client_ids}
columns = ["qty", "shipping_price", "total_price", "line_cost", "line_profit"]
for column in columns:
    data_dict[column] = [client_spend(client_id, column) for client_id in top_client_ids]

summary_df = pd.DataFrame(data_dict).sort_values("line_profit", ascending=False)
summary_df


client_id     qty  shipping_price  total_price    line_cost  line_profit
4      24741  239862      5126448.37  82268891.98  45688899.71  36579992.27
3      38378   73667      3429455.40  12906550.88   9634720.98   3271829.90
1      66037   43018      1395151.85  10259514.80   7004482.98   3255031.82
2      46820   75768      1601448.84   9743794.32   7007191.64   2736602.68
0      33615   64313      1828984.89   8377308.57   6175313.91   2201994.66

In [48]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
money_columns = ['line_cost', 'shipping_price', 'total_price', 'line_profit']

def format_money(amount):
    return f"${amount/1000000:.2f}M"

for column in money_columns:
    summary_df[column] = summary_df[column].apply(format_money)

rename_columns = {
    'qty': 'Total Units Purchased',
    'shipping_price': 'Total Shipping Price',
    'total_price': 'Total Revenue',
    'line_profit': 'Total Profit',
    'client_id': 'Client ID',
    'line_cost': 'Total Cost'
}
summary_df = summary_df.rename(columns=rename_columns)


In [49]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df.sort_values('Total Profit', ascending=False)

Client ID  Total Units Purchased Total Shipping Price Total Revenue  \
4      24741                 239862               $5.13M       $82.27M   
3      38378                  73667               $3.43M       $12.91M   
1      66037                  43018               $1.40M       $10.26M   
2      46820                  75768               $1.60M        $9.74M   
0      33615                  64313               $1.83M        $8.38M   

  Total Cost Total Profit  
4    $45.69M      $36.58M  
3     $9.63M       $3.27M  
1     $7.00M       $3.26M  
2     $7.01M       $2.74M  
0     $6.18M       $2.20M

In [50]:
In this exercise, we were able to use the data to find that the top three categories of products were, Consumables, Furniture & Software. 
By adding new columns, Subtotal, Shipping price, Total price, Line cost, & Line profit, we were able to calculate the spend amount, total units purchased, total shipping price, 
total revenue & total profit for the top 5 clients. 

SyntaxError: invalid syntax (1738742342.py, line 1)